samza-core/src/main/scala/org/apache/samza/coordinator/JobModelManager.scala

```diff
@@ -200,22 +299,18 @@ object JobModelManager extends Logging {
     // SSPTaskNameGrouper for locality, load-balancing, etc.
     val containerGrouperFactory = Util.getObj(config.getTaskNameGrouperFactory, classOf[TaskNameGrouperFactory])
     val containerGrouper = containerGrouperFactory.build(config)
-    val containerModels = {
-      containerGrouper match {
-        case grouper: BalancingTaskNameGrouper if isHostAffinityEnabled => grouper.balance(taskModels.asJava, localityManager)
-        case _ => containerGrouper.group(taskModels.asJava, containerIds)
-      }
-    }
-    val containerMap = containerModels.asScala.map { case (containerModel) => containerModel.getId -> containerModel }.toMap
-
-    if (isHostAffinityEnabled) {
-      new JobModel(config, containerMap.asJava, localityManager)
+    var containerModels: util.Set[ContainerModel] = null
+    if(isHostAffinityEnabled) {
+      containerModels = containerGrouper.group(taskModels, grouperContext)
```

Comment:
It would be much cleaner to have a single path here and let the grouper decide on how it wants to perform grouping? I think host affinity is a detail that shouldn't determine which grouping choice to do in the JobModelManager since that is sort of leaking the grouper responsibility into JobModelManager.

IMO, it is cleaner to have the JobModelManager prepare the context for the grouper and hand it off to the grouper which in turn determines how to perform grouping based on the context it has.

Let me know what you think?